File: sqlparser/src/sql/NormalizationOption.java
//@
package cdata.sql;
import rssbus.oputils.common.*;
import java.util.*;
//@
/*#
using RSSBus.core;
using RSSBus;
namespace CData.Sql {
#*/
public final class NormalizationOption {
  private final boolean _implicitCrossJoin;
  private final boolean _implicitNaturalJoin;
  private final boolean _tableExprWithQuery;
  private final boolean _criteriaWithSubQuery;
  private final boolean _criteriaWithNot;
  private final boolean _criteria;
  private final boolean _resolveTableAlias;
  private final boolean _fixUniqueAlias;
  private final boolean _criteriaInJoin;
  private final boolean _implicitCommaJoin;
  private final boolean _implicitInnerJoin;
  private final boolean _operationExpression;
  private final boolean _appendNestedQueryTableName;
  private final boolean _rightJoin;
  private final boolean _minimizeCriteria;
  private final boolean _constantColumn;
  private final boolean _formulaAlias;
  private final boolean _nestedJoin;
  private final boolean _distinct;
  private final boolean _count_distinct;
  private final boolean _predictTrue;
  private final boolean _equiInnerJoin;
  private final boolean _semi_anti_join;
  private final boolean _functionSubstitute;
  private final boolean _removeDistinctIfColumnUnique;
  
  public static final String[] DEFAULT_NORMALIZATIONS  = new String[] {
          "ImplicitNaturalJoin",
          "TableExprWithQuery",
          "ConstantColumn",
          "Distinct",
          "CriteriaWithSubQuery",
          "CriteriaWithNot",
          "Criteria",
          "MinimizeCriteria",
          "ResolveTableAlias",
          "FixUniqueAlias",
          "CriteriaInJoin",
          "OperationExpression",
          "AppendNestedQueryTableName",
          "NestedJoin",
          "PredictTrue",
          "EquiInnerJoin",
          "FunctionSubstitute"
  };

  public static NormalizationOption buildNormalizationOption(String[] normalizationOptions) {
    NormalizationOptionBuilder builder = new NormalizationOptionBuilder();
    if (normalizationOptions != null && normalizationOptions.length > 0) {
      for (String opt : normalizationOptions) {
        if (opt.equalsIgnoreCase("ImplicitNaturalJoin")) {
          builder.config(NormalizationOptions.ImplicitNaturalJoin, true);
        } else if (opt.equalsIgnoreCase("ImplicitCrossJoin")) {
          builder.config(NormalizationOptions.ImplicitCrossJoin, true);
        } else if (opt.equalsIgnoreCase("ImplicitCommaJoin")) {
          builder.config(NormalizationOptions.ImplicitCommaJoin, true);
        }  else if (opt.equalsIgnoreCase("ImplicitInnerJoin")) {
          builder.config(NormalizationOptions.ImplicitInnerJoin, true);
        } else if (opt.equalsIgnoreCase("TableExprWithQuery")) {
          builder.config(NormalizationOptions.TableExprWithQuery, true);
        }  else if (opt.equalsIgnoreCase("ConstantColumn")) {
          builder.config(NormalizationOptions.ConstantColumn, true);
        } else if (opt.equalsIgnoreCase("CriteriaWithSubQuery")) {
          builder.config(NormalizationOptions.CriteriaWithSubQuery, true);
        } else if (opt.equalsIgnoreCase("CriteriaWithNot")) {
          builder.config(NormalizationOptions.CriteriaWithNot, true);
        } else if (opt.equalsIgnoreCase("Criteria")) {
          builder.config(NormalizationOptions.Criteria, true);
        } else if (opt.equalsIgnoreCase("ResolveTableAlias")) {
          builder.config(NormalizationOptions.ResolveTableAlias, true);
        } else if (opt.equalsIgnoreCase("FixUniqueAlias")) {
          builder.config(NormalizationOptions.FixUniqueAlias, true);
        } else if (opt.equalsIgnoreCase("CriteriaInJoin")) {
          builder.config(NormalizationOptions.CriteriaInJoin, true);
        } else if (opt.equalsIgnoreCase("OperationExpression")) {
          builder.config(NormalizationOptions.OperationExpression, true);
        } else if (opt.equalsIgnoreCase("AppendNestedQueryTableName")) {
          builder.config(NormalizationOptions.AppendNestedQueryTableName, true);
        } else if (opt.equalsIgnoreCase("RightJoin")) {
          builder.config(NormalizationOptions.RightJoin, true);
        } else if (opt.equalsIgnoreCase("MinimizeCriteria")) {
          builder.config(NormalizationOptions.MinimizeCriteria, true);
        } else if (opt.equalsIgnoreCase("FormulaAlias")) {
          builder.config(NormalizationOptions.FormulaAlias, true);
        } else if (opt.equalsIgnoreCase("NestedJoin")) {
          builder.config(NormalizationOptions.NestedJoin, true);
        } else if (opt.equalsIgnoreCase("Distinct")) {
          builder.config(NormalizationOptions.Distinct, true);
        } else if (opt.equalsIgnoreCase("Count_Distinct")) {
          builder.config(NormalizationOptions.Count_Distinct, true);
        } else if (opt.equalsIgnoreCase("PredictTrue")) {
          builder.config(NormalizationOptions.PredictTrue, true);
        } else if (opt.equalsIgnoreCase("EquiInnerJoin")) {
          builder.config(NormalizationOptions.EquiInnerJoin, true);
        } else if (opt.equalsIgnoreCase("SemiAntiJoin")) {
          builder.config(NormalizationOptions.SemiAntiJoin, true);
        } else if (opt.equalsIgnoreCase("FunctionSubstitute")) {
          builder.config(NormalizationOptions.FunctionSubstitute, true);
        } else if (opt.equalsIgnoreCase("RemoveDistinctIfColumnUnique")) {
          builder.config(NormalizationOptions.RemoveDistinctIfColumnUnique, true);
        }
      }
    }
    builder.config(NormalizationOptions.FormulaAlias, true);
    builder.config(NormalizationOptions.ResolveTableAlias, true);
    return builder.build();
  }

  public static final String[] setImplicitJoin(String [] DEFAULT_NORMALIZATIONS, String implicitJoinType) {
    String[] result = DEFAULT_NORMALIZATIONS;
    String temp = Utilities.toUpperCase(implicitJoinType);
    if (temp != null && temp.startsWith("IMPLICIT")) {
      result = new String[DEFAULT_NORMALIZATIONS.length];
      for (int i = 0 ; i < DEFAULT_NORMALIZATIONS.length; ++i) {
        String optionName = Utilities.toUpperCase(DEFAULT_NORMALIZATIONS[i]);
        if (optionName.startsWith("IMPLICIT")) {
          result[i] = implicitJoinType;
        } else {
          result[i] = optionName;
        }
      }
    }
    return result;
  }

  public static final String[] AddToDefaultNormalizations(String normalizations) {
    Vector<String> modifiedOptions = new Vector<String>();
    if(normalizations != null) {
      String[] add= Utilities.splitString(normalizations.toUpperCase(),",");
      int total = DEFAULT_NORMALIZATIONS.length;
      for(int i=0;i<DEFAULT_NORMALIZATIONS.length;i++){
        modifiedOptions.add(DEFAULT_NORMALIZATIONS[i].toUpperCase());
      }
      for(String s : add) {
        if(s.startsWith("-")) modifiedOptions.remove(s.substring(1));
        else if(s.startsWith("+")) modifiedOptions.add(s.substring(1));
        else modifiedOptions.add(s);
      }
    }
    String[] retValue = new String[modifiedOptions.size()];
    for(int i =0; i<modifiedOptions.size(); i++) {
      retValue[i] = modifiedOptions.elementAt(i);
    }
    return retValue;
  }

  public static final String[] addToNormalizations(String [] normalizations, String addings) {
    String[] add = Utilities.splitString(addings, ",");

    if (null == normalizations) {
      normalizations = new String[0];
    }
    int total = normalizations.length + add.length;
    String[] result = new String[total];
    for(int i=0;i<normalizations.length;i++){
      result[i] = normalizations[i];
    }
    for(int i=normalizations.length, j=0;i<total;i++, j++){
      result[i] = add[j];
    }
    return result;
  }

  public NormalizationOption(NormalizationOptionBuilder builder) {
    this._implicitCommaJoin = builder.getConfigAsBoolean(NormalizationOptions.ImplicitCommaJoin);
    this._implicitNaturalJoin = builder.getConfigAsBoolean(NormalizationOptions.ImplicitNaturalJoin);
    this._implicitCrossJoin = builder.getConfigAsBoolean(NormalizationOptions.ImplicitCrossJoin);
    this._implicitInnerJoin = builder.getConfigAsBoolean(NormalizationOptions.ImplicitInnerJoin);
    this._tableExprWithQuery = builder.getConfigAsBoolean(NormalizationOptions.TableExprWithQuery);
    this._criteriaWithSubQuery = builder.getConfigAsBoolean(NormalizationOptions.CriteriaWithSubQuery);
    this._criteriaWithNot = builder.getConfigAsBoolean(NormalizationOptions.CriteriaWithNot);
    this._criteria = builder.getConfigAsBoolean(NormalizationOptions.Criteria);
    this._resolveTableAlias = builder.getConfigAsBoolean(NormalizationOptions.ResolveTableAlias);
    this._fixUniqueAlias = builder.getConfigAsBoolean(NormalizationOptions.FixUniqueAlias);
    this._criteriaInJoin = builder.getConfigAsBoolean(NormalizationOptions.CriteriaInJoin);
    this._operationExpression = builder.getConfigAsBoolean(NormalizationOptions.OperationExpression);
    this._appendNestedQueryTableName = builder.getConfigAsBoolean(NormalizationOptions.AppendNestedQueryTableName);
    this._rightJoin = builder.getConfigAsBoolean(NormalizationOptions.RightJoin);
    this._minimizeCriteria = builder.getConfigAsBoolean(NormalizationOptions.MinimizeCriteria);
    this._constantColumn = builder.getConfigAsBoolean(NormalizationOptions.ConstantColumn);
    this._formulaAlias = builder.getConfigAsBoolean(NormalizationOptions.FormulaAlias);
    this._nestedJoin = builder.getConfigAsBoolean(NormalizationOptions.NestedJoin);
    this._distinct = builder.getConfigAsBoolean(NormalizationOptions.Distinct);
    this._count_distinct = builder.getConfigAsBoolean(NormalizationOptions.Count_Distinct);
    this._predictTrue = builder.getConfigAsBoolean(NormalizationOptions.PredictTrue);
    this._equiInnerJoin = builder.getConfigAsBoolean(NormalizationOptions.EquiInnerJoin);
    this._semi_anti_join = builder.getConfigAsBoolean(NormalizationOptions.SemiAntiJoin);
    this._functionSubstitute = builder.getConfigAsBoolean(NormalizationOptions.FunctionSubstitute);
    this._removeDistinctIfColumnUnique = builder.getConfigAsBoolean(NormalizationOptions.RemoveDistinctIfColumnUnique);
  }

  public boolean normalizeImplicitCommaJoin() {
    return this._implicitCommaJoin;
  }

  public boolean normalizeImplicitCrossJoin() {
    return this._implicitCrossJoin;
  }

  public boolean normalizeImplicitInnerJoin() {
    return this._implicitInnerJoin;
  }

  public boolean normalizeImplicitNaturalJoin() {
    return this._implicitNaturalJoin;
  }

  public boolean normalizeTableExprWithQuery() {
    return this._tableExprWithQuery;
  }

  public boolean normalizeCriteriaWithSubQuery() {
    return _criteriaWithSubQuery;
  }

  public boolean normalizeCriteriaWithNot() {
    return _criteriaWithNot;
  }

  public boolean normalizeResolveTableAlias() {
    return _resolveTableAlias;
  }

  public boolean normalizeFixUniqueAlias() {
    return this._fixUniqueAlias;
  }

  public boolean normalizeCriteria() {
    return this._criteria;
  }

  public boolean normalizeCriteriaInJoin() {
    return this._criteriaInJoin;
  }

  public boolean normalizeOperationExpression() {
    return this._operationExpression;
  }

  public boolean normalizeNestedQueryTableName() {
    return this._appendNestedQueryTableName;
  }

  public boolean normalizeRightJoin() {
    return this._rightJoin;
  }

  public boolean normalizeMinimizeCriteria() {
    return this._minimizeCriteria;
  }

  public boolean normalizeConstantColumn() {
    return this._constantColumn;
  }

  public boolean normalizeFormulaAlias() {
    return this._formulaAlias;
  }

  public boolean normalizeNestedJoin() {
    return this._nestedJoin;
  }

  public boolean normalizeDistinct() {
    return this._distinct;
  }

  public boolean normalizeCountDistinct() {
    return this._count_distinct;
  }

  public boolean normalizePredictTrue() {
    return this._predictTrue;
  }

  public boolean normalizeEquiInnerJoin() {
    return this._equiInnerJoin;
  }

  public boolean normalizedSemiAntiJoin() {
    return this._semi_anti_join;
  }

  public boolean normalizedFunctionSubstitute() {
    return this._functionSubstitute;
  }

  public boolean normalizeRemoveDistinctIfColumnUnique() {
    return this._removeDistinctIfColumnUnique;
  }
}

File: sqlparser/src/sql/csgen/NormalizationOption.cs
using System;
using System.Collections;
using System.IO;
using System.Threading;
using System.Net;
using System.Net.Sockets;
using System.Text;
using RSSBus.core.j2cs;



using RSSBus.core;
using RSSBus;
namespace CData.Sql {

public sealed class NormalizationOption {
  private readonly bool _implicitCrossJoin;
  private readonly bool _implicitNaturalJoin;
  private readonly bool _tableExprWithQuery;
  private readonly bool _criteriaWithSubQuery;
  private readonly bool _criteriaWithNot;
  private readonly bool _criteria;
  private readonly bool _resolveTableAlias;
  private readonly bool _fixUniqueAlias;
  private readonly bool _criteriaInJoin;
  private readonly bool _implicitCommaJoin;
  private readonly bool _implicitInnerJoin;
  private readonly bool _operationExpression;
  private readonly bool _appendNestedQueryTableName;
  private readonly bool _rightJoin;
  private readonly bool _minimizeCriteria;
  private readonly bool _constantColumn;
  private readonly bool _formulaAlias;
  private readonly bool _nestedJoin;
  private readonly bool _distinct;
  private readonly bool _count_distinct;
  private readonly bool _predictTrue;
  private readonly bool _equiInnerJoin;
  private readonly bool _semi_anti_join;
  private readonly bool _functionSubstitute;
  private readonly bool _removeDistinctIfColumnUnique;
  
  public static string[] DEFAULT_NORMALIZATIONS  = new string[] {
          "ImplicitNaturalJoin",
          "TableExprWithQuery",
          "ConstantColumn",
          "Distinct",
          "CriteriaWithSubQuery",
          "CriteriaWithNot",
          "Criteria",
          "MinimizeCriteria",
          "ResolveTableAlias",
          "FixUniqueAlias",
          "CriteriaInJoin",
          "OperationExpression",
          "AppendNestedQueryTableName",
          "NestedJoin",
          "PredictTrue",
          "EquiInnerJoin",
          "FunctionSubstitute"
  };

  public static NormalizationOption BuildNormalizationOption(string[] normalizationOptions) {
    NormalizationOptionBuilder builder = new NormalizationOptionBuilder();
    if (normalizationOptions != null && normalizationOptions.Length > 0) {
      foreach(String opt in normalizationOptions) {
        if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "ImplicitNaturalJoin")) {
          builder.Config(NormalizationOptions.ImplicitNaturalJoin, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "ImplicitCrossJoin")) {
          builder.Config(NormalizationOptions.ImplicitCrossJoin, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "ImplicitCommaJoin")) {
          builder.Config(NormalizationOptions.ImplicitCommaJoin, true);
        }  else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "ImplicitInnerJoin")) {
          builder.Config(NormalizationOptions.ImplicitInnerJoin, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "TableExprWithQuery")) {
          builder.Config(NormalizationOptions.TableExprWithQuery, true);
        }  else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "ConstantColumn")) {
          builder.Config(NormalizationOptions.ConstantColumn, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "CriteriaWithSubQuery")) {
          builder.Config(NormalizationOptions.CriteriaWithSubQuery, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "CriteriaWithNot")) {
          builder.Config(NormalizationOptions.CriteriaWithNot, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "Criteria")) {
          builder.Config(NormalizationOptions.Criteria, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "ResolveTableAlias")) {
          builder.Config(NormalizationOptions.ResolveTableAlias, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "FixUniqueAlias")) {
          builder.Config(NormalizationOptions.FixUniqueAlias, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "CriteriaInJoin")) {
          builder.Config(NormalizationOptions.CriteriaInJoin, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "OperationExpression")) {
          builder.Config(NormalizationOptions.OperationExpression, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "AppendNestedQueryTableName")) {
          builder.Config(NormalizationOptions.AppendNestedQueryTableName, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "RightJoin")) {
          builder.Config(NormalizationOptions.RightJoin, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "MinimizeCriteria")) {
          builder.Config(NormalizationOptions.MinimizeCriteria, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "FormulaAlias")) {
          builder.Config(NormalizationOptions.FormulaAlias, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "NestedJoin")) {
          builder.Config(NormalizationOptions.NestedJoin, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "Distinct")) {
          builder.Config(NormalizationOptions.Distinct, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "Count_Distinct")) {
          builder.Config(NormalizationOptions.Count_Distinct, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "PredictTrue")) {
          builder.Config(NormalizationOptions.PredictTrue, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "EquiInnerJoin")) {
          builder.Config(NormalizationOptions.EquiInnerJoin, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "SemiAntiJoin")) {
          builder.Config(NormalizationOptions.SemiAntiJoin, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "FunctionSubstitute")) {
          builder.Config(NormalizationOptions.FunctionSubstitute, true);
        } else if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(opt, "RemoveDistinctIfColumnUnique")) {
          builder.Config(NormalizationOptions.RemoveDistinctIfColumnUnique, true);
        }
      }
    }
    builder.Config(NormalizationOptions.FormulaAlias, true);
    builder.Config(NormalizationOptions.ResolveTableAlias, true);
    return builder.Build();
  }

  public static string[] SetImplicitJoin(string [] DEFAULT_NORMALIZATIONS, string implicitJoinType) {
    string[] result = DEFAULT_NORMALIZATIONS;
    string temp = Utilities.ToUpper(implicitJoinType);
    if (temp != null && temp.StartsWith("IMPLICIT")) {
      result = new string[DEFAULT_NORMALIZATIONS.Length];
      for (int i = 0 ; i < DEFAULT_NORMALIZATIONS.Length; ++i) {
        string optionName = Utilities.ToUpper(DEFAULT_NORMALIZATIONS[i]);
        if (optionName.StartsWith("IMPLICIT")) {
          result[i] = implicitJoinType;
        } else {
          result[i] = optionName;
        }
      }
    }
    return result;
  }

  public static string[] AddToDefaultNormalizations(string normalizations) {
    JavaVector<string> modifiedOptions = new JavaVector<string>();
    if(normalizations != null) {
      string[] add= Utilities.SplitString(normalizations.ToUpper(),",");
      int total = DEFAULT_NORMALIZATIONS.Length;
      for(int i=0;i<DEFAULT_NORMALIZATIONS.Length;i++){
        modifiedOptions.Add(DEFAULT_NORMALIZATIONS[i].ToUpper());
      }
      foreach(String s in add) {
        if(s.StartsWith("-")) modifiedOptions.Remove(RSSBus.core.j2cs.Converter.GetSubstring(s, 1));
        else if(s.StartsWith("+")) modifiedOptions.Add(RSSBus.core.j2cs.Converter.GetSubstring(s, 1));
        else modifiedOptions.Add(s);
      }
    }
    string[] retValue = new string[modifiedOptions.Size()];
    for(int i =0; i<modifiedOptions.Size(); i++) {
      retValue[i] = modifiedOptions.ElementAt(i);
    }
    return retValue;
  }

  public static string[] AddToNormalizations(string [] normalizations, string addings) {
    string[] add = Utilities.SplitString(addings, ",");

    if (null == normalizations) {
      normalizations = new string[0];
    }
    int total = normalizations.Length + add.Length;
    string[] result = new string[total];
    for(int i=0;i<normalizations.Length;i++){
      result[i] = normalizations[i];
    }
    for(int i=normalizations.Length, j=0;i<total;i++, j++){
      result[i] = add[j];
    }
    return result;
  }

  public NormalizationOption(NormalizationOptionBuilder builder) {
    this._implicitCommaJoin = builder.GetConfigAsBoolean(NormalizationOptions.ImplicitCommaJoin);
    this._implicitNaturalJoin = builder.GetConfigAsBoolean(NormalizationOptions.ImplicitNaturalJoin);
    this._implicitCrossJoin = builder.GetConfigAsBoolean(NormalizationOptions.ImplicitCrossJoin);
    this._implicitInnerJoin = builder.GetConfigAsBoolean(NormalizationOptions.ImplicitInnerJoin);
    this._tableExprWithQuery = builder.GetConfigAsBoolean(NormalizationOptions.TableExprWithQuery);
    this._criteriaWithSubQuery = builder.GetConfigAsBoolean(NormalizationOptions.CriteriaWithSubQuery);
    this._criteriaWithNot = builder.GetConfigAsBoolean(NormalizationOptions.CriteriaWithNot);
    this._criteria = builder.GetConfigAsBoolean(NormalizationOptions.Criteria);
    this._resolveTableAlias = builder.GetConfigAsBoolean(NormalizationOptions.ResolveTableAlias);
    this._fixUniqueAlias = builder.GetConfigAsBoolean(NormalizationOptions.FixUniqueAlias);
    this._criteriaInJoin = builder.GetConfigAsBoolean(NormalizationOptions.CriteriaInJoin);
    this._operationExpression = builder.GetConfigAsBoolean(NormalizationOptions.OperationExpression);
    this._appendNestedQueryTableName = builder.GetConfigAsBoolean(NormalizationOptions.AppendNestedQueryTableName);
    this._rightJoin = builder.GetConfigAsBoolean(NormalizationOptions.RightJoin);
    this._minimizeCriteria = builder.GetConfigAsBoolean(NormalizationOptions.MinimizeCriteria);
    this._constantColumn = builder.GetConfigAsBoolean(NormalizationOptions.ConstantColumn);
    this._formulaAlias = builder.GetConfigAsBoolean(NormalizationOptions.FormulaAlias);
    this._nestedJoin = builder.GetConfigAsBoolean(NormalizationOptions.NestedJoin);
    this._distinct = builder.GetConfigAsBoolean(NormalizationOptions.Distinct);
    this._count_distinct = builder.GetConfigAsBoolean(NormalizationOptions.Count_Distinct);
    this._predictTrue = builder.GetConfigAsBoolean(NormalizationOptions.PredictTrue);
    this._equiInnerJoin = builder.GetConfigAsBoolean(NormalizationOptions.EquiInnerJoin);
    this._semi_anti_join = builder.GetConfigAsBoolean(NormalizationOptions.SemiAntiJoin);
    this._functionSubstitute = builder.GetConfigAsBoolean(NormalizationOptions.FunctionSubstitute);
    this._removeDistinctIfColumnUnique = builder.GetConfigAsBoolean(NormalizationOptions.RemoveDistinctIfColumnUnique);
  }

  public bool NormalizeImplicitCommaJoin() {
    return this._implicitCommaJoin;
  }

  public bool NormalizeImplicitCrossJoin() {
    return this._implicitCrossJoin;
  }

  public bool NormalizeImplicitInnerJoin() {
    return this._implicitInnerJoin;
  }

  public bool NormalizeImplicitNaturalJoin() {
    return this._implicitNaturalJoin;
  }

  public bool NormalizeTableExprWithQuery() {
    return this._tableExprWithQuery;
  }

  public bool NormalizeCriteriaWithSubQuery() {
    return _criteriaWithSubQuery;
  }

  public bool NormalizeCriteriaWithNot() {
    return _criteriaWithNot;
  }

  public bool NormalizeResolveTableAlias() {
    return _resolveTableAlias;
  }

  public bool NormalizeFixUniqueAlias() {
    return this._fixUniqueAlias;
  }

  public bool NormalizeCriteria() {
    return this._criteria;
  }

  public bool NormalizeCriteriaInJoin() {
    return this._criteriaInJoin;
  }

  public bool NormalizeOperationExpression() {
    return this._operationExpression;
  }

  public bool NormalizeNestedQueryTableName() {
    return this._appendNestedQueryTableName;
  }

  public bool NormalizeRightJoin() {
    return this._rightJoin;
  }

  public bool NormalizeMinimizeCriteria() {
    return this._minimizeCriteria;
  }

  public bool NormalizeConstantColumn() {
    return this._constantColumn;
  }

  public bool NormalizeFormulaAlias() {
    return this._formulaAlias;
  }

  public bool NormalizeNestedJoin() {
    return this._nestedJoin;
  }

  public bool NormalizeDistinct() {
    return this._distinct;
  }

  public bool NormalizeCountDistinct() {
    return this._count_distinct;
  }

  public bool NormalizePredictTrue() {
    return this._predictTrue;
  }

  public bool NormalizeEquiInnerJoin() {
    return this._equiInnerJoin;
  }

  public bool NormalizedSemiAntiJoin() {
    return this._semi_anti_join;
  }

  public bool NormalizedFunctionSubstitute() {
    return this._functionSubstitute;
  }

  public bool NormalizeRemoveDistinctIfColumnUnique() {
    return this._removeDistinctIfColumnUnique;
  }
}
}

